### PR TITLE
Android: Support games in ZIP archives

### LINKED ISSUE #2760: Android: Support games in ZIP archives
Just an issue to track this.

The Android Player should be able to detect whether a game is in a ZIP archive.

For this the Game Browser should at least:

- Parse the ZIP and look for RPG_RT.ldb & RPG_RT.lmt
- Pass this proper ZIP path to the Player on launch
- Set a proper Savegame directory
- (Listing the title is likely too slow as this requires decompressing first? Lets think about this as an additional feature later)

### PR DESCRIPTION
**Pull request by Ghabry**
Fix #2760

```
[2023-01-27 14:25:08] Debug: CLI: --project-path content://com.android.externalstorage.documents/tree/primary%3Aeasyrpg/document/primary%3Aeasyrpg%2Fgames%2FDon's%20Adventures.zip/Deep/Don's Adventures --save-path content://com.android.externalstorage.documents/tree/primary%3Aeasyrpg/document/primary%3Aeasyrpg%2Fsaves%2FDon's%20Adventures
[...]
[2023-01-27 14:25:09] Debug: Detected encoding: ibm-5348_P100-1997
[2023-01-27 14:25:09] Debug: Game Directory:
[2023-01-27 14:25:09] Debug: 1: [Zip] com.android.externalstorage.documents/tree/primary%3Aeasyrpg/document/primary%3Aeasyrpg%2Fgames%2FDon's%20Adventures.zip (UTF-8) -> Deep/Don's Adventures
[2023-01-27 14:25:09] Debug: SaveDirectory:
[2023-01-27 14:25:09] Debug: 1: [SAF]  -> com.android.externalstorage.documents/tree/primary%3Aeasyrpg/document/primary%3Aeasyrpg%2Fsaves%2FDon's%20Adventures
[2023-01-27 14:25:09] Debug: Loading game Don's Adventures
```

Limitations:

- ~~Setting of the encoding writes directly in the ini and is unsupported~~
- ~~No title graphic shown~~
- All the information should be cached as the startup is a bit slow (but still acceptable)

Not so relevant limitations:

- Only the first game in the ZIP is returned
- No support for ZIP inside ZIP

~~TODO: Update the "Getting started" message~~

Screenshot_2023-01-27-14-29-48-412_org easyrpg player

### CODE REVIEW

File: builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameScanner.java [hunk 1]
```diff
@@ -121,6 +131,12 @@ else if (depth > 1) {
                             // Android SAF calls and scanFolder(...) already check that)
                             scanFolderRecursive(context, fileURI, depth - 1);
                         }
+                    } else if (fileDocumentID.endsWith(".zip")) {
```
Comment (fdelapena) [inline builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameScanner.java:134]: Maybe is cheap adding .easyrpg extension as well

**Comment by Ghabry** [comment]
Encoding for games in ZIP archives is now supported. Had to remove the "parse RPG_RT.ini" part though because writing is impossible. Instead uses shared preferences.

Title images from ZIP archives are now displayed.

Making this PR ready. Implementing a cache is a task for another day :p.

Screenshot_2023-03-08-23-42-57-969_org easyrpg player

**Review by fdelapena — APPROVED** [review]
fileDocumentID.toLowerCase().endsWith() could be useful.

**Comment by Ghabry** [comment]
Well. Here is the cache. THAT SPEEEEEEEEEEEED! (on 2nd startup)

The cache calculates a "hash" of the main games folder (hashcode of names + type ). When the hash is the same the cache is returned. Otherwise it does a rescan. Can of course miss some rescans when only subfolders are modified but there is still this "force rescan" button available.

**Review by fdelapena — APPROVED** [review]

**Comment by fdelapena** [comment]
jenkins: test this please

**Pull request merged**